monero-sys/src/lib.rs

```diff
@@ -0,0 +1,1511 @@
+    }
+
+    /// Get the main address of the wallet.
+    /// The main address is the first address of the first account.
+    pub async fn main_address(&self) -> monero::Address {
+        self.call(move |wallet| wallet.main_address()).await
+    }
+
+    /// Get the current height of the blockchain.
+    /// May involve an RPC call to the daemon.
+    /// Returns `None` if the wallet is not connected to a daemon.
+    ///
+    /// Retries at most 5 times with a 500ms delay between attempts.
+    pub async fn blockchain_height(&self) -> anyhow::Result<u64> {
+        const MAX_RETRIES: u64 = 5;
+
+        for _ in 0..MAX_RETRIES {
+            if let Some(height) = self
+                .call(move |wallet| wallet.daemon_blockchain_height())
+                .await
+            {
+                return Ok(height);
+            }
+
+            tokio::time::sleep(std::time::Duration::from_millis(500)).await;
+        }
+
+        self.check_wallet().await?;
+
+        bail!("Failed to get blockchain height after 5 attempts");
+    }
+
+    /// Transfer funds to an address.
+    pub async fn transfer(
+        &self,
+        address: &monero::Address,
+        amount: monero::Amount,
+    ) -> anyhow::Result<TxReceipt> {
+        let address = *address;
+        self.call(move |wallet| wallet.transfer(&address, amount))
+            .await
+    }
+
+    /// Sweep all funds to an address.
+    pub async fn sweep(&self, address: &monero::Address) -> anyhow::Result<Vec<String>> {
+        let address = *address;
+        self.call(move |wallet| wallet.sweep(&address)).await
+    }
+
+    /// Get the seed of the wallet.
+    pub async fn seed(&self) -> String {
+        self.call(move |wallet| wallet.seed()).await
+    }
+
+    /// Get the creation height of the wallet.
+    pub async fn creation_height(&self) -> u64 {
+        self.call(move |wallet| wallet.creation_height()).await
+    }
+
+    /// Get the unlocked balance of the wallet.
+    pub async fn unlocked_balance(&self) -> monero::Amount {
+        self.call(move |wallet| wallet.unlocked_balance()).await
+    }
+
+    /// Get the total balance of the wallet.
+    pub async fn total_balance(&self) -> monero::Amount {
+        self.call(move |wallet| wallet.total_balance()).await
+    }
+
+    /// Check if the wallet is synchronized.
+    async fn synchronized(&self) -> bool {
+        self.call(move |wallet| wallet.synchronized()).await
+    }
+
+    /// Get the sync progress of the wallet.
+    async fn sync_progress(&self) -> SyncProgress {
+        self.call(move |wallet| wallet.sync_progress()).await
+    }
+
+    /// Check if the wallet is connected to a daemon.
+    pub async fn connected(&self) -> bool {
+        self.call(move |wallet| wallet.connected()).await
+    }
+
+    /// Check that the wallet is created and ready to use.
+    /// Call this after creating a wallet to make sure the wallet thread responds correctly.
+    async fn check_wallet(&self) -> anyhow::Result<()> {
```

Comment:
How come this is called `check_wallet` even though it doesn't do anything wallet specific

Reply:
this function verifies the channel connection to the wallet thread. it's needed to verify the success of the wallet initialization because that has to happen in the thread